src/github/github.installation.service.ts

```diff
@@ -0,0 +1,173 @@
+    });
+    return !!org;
+  }
+
+  async isUserAllowed(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<boolean> {
+    try {
+      const org = await this.findByName(organizationName);
+      if (org.status !== OrganizationStatus.ACTIVE) {
+        return false;
+      }
+      if (org.allowAllUsers) {
+        return true;
+      }
+      return org.activeUsers.includes(githubUsername);
+    } catch (error) {
+      this.logger.error(`Error checking user allowance: ${error.message}`);
+      return false;
+    }
+  }
+
+  async addAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    if (org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers = org.inactiveUsers.filter(
+        (username) => username !== githubUsername,
+      );
+    }
+    if (!org.activeUsers.includes(githubUsername)) {
+      org.activeUsers.push(githubUsername);
+      return this.allowedOrgRepository.save(org);
+    }
+    return org;
+  }
+
+  async removeAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.activeUsers = org.activeUsers.filter(
+      (username) => username !== githubUsername,
+    );
+    if (!org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers.push(githubUsername);
+    }
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async updateStatus(
+    organizationName: string,
+    newStatus: OrganizationStatus,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.status = newStatus;
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async uninstall(id: number) {
+    const now = Math.floor(Date.now() / 1000);
+
+    const pemFilename = Env.GITHUB.APP_PEM;
+    const pemFilePath = path.join(__dirname, '../../../', pemFilename);
+    const privateKey = fs.readFileSync(pemFilePath, 'utf8');
+
+    const payload = {
+      iat: now,
+      exp: now + 600,
+      iss: Env.GITHUB.APP_ID,
+      installation_id: id,
+    };
+    const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });
+    const octokit = new Octokit({
+      auth: token,
+    });
+
+    await octokit.request('DELETE /app/installations/{installation_id}', {
+      installation_id: id,
+      headers: {
+        'X-GitHub-Api-Version': '2022-11-28',
+      },
+    });
+  }
+
+  async allowedOrganisations(name: string) {
+    return this.allowedOrgRepository.find({ where: { adminUserName: name } });
+  }
```

Comment:
PullProofAI

Consider renaming the `allowedOrganisations` method to `getOrganizationsByAdminUserName` for clarity on its purpose.

```diff
+ // Example of potential improvement for create method validation
+ async create(createOrgDto: CreateOrgDto): Promise<AllowedOrganization> {
+     if (!createOrgDto.name || !createOrgDto.adminUserName) {
+         throw new Error('Name and admin username are required for organization creation');
+     }
+     const newOrg = this.allowedOrgRepository.create(createOrgDto);
+     return this.allowedOrgRepository.save(newOrg);
+ }
```

```diff
+ // Example of potential improvement for findOne method error handling
+ async findOne(id: string): Promise<AllowedOrganization> {
+     try {
+         const org = await this.allowedOrgRepository.findOne({ where: { id } });
+         if (!org) {
+             throw new NotFoundException(`Organization not found`);
+         }
+         return org;
+     } catch (error) {
+         throw new Error(`Error finding organization: ${error.message}`);
+     }
+ }
```

```diff
+ // Example of potential improvement for isOrganizationAllowed method logging
+ async isOrganizationAllowed(organizationName: string): Promise<boolean> {
+     try {
+         const org = await this.allowedOrgRepository.findOne({
+             where: { organizationName, status: OrganizationStatus.ACTIVE },
+         });
+         return !!org;
+     } catch (error) {
+         this.logger.error(`Error checking organization allowance: ${error.message}`);
+         return false;
+     }
+ }
```



<!-- This is an auto-generated comment by PullProofAI -->